File: payment-service/src/main/java/com/hedvig/paymentservice/services/bankAccounts/BankAccountServiceImpl.java
package com.hedvig.paymentservice.services.bankAccounts;

import com.hedvig.paymentservice.domain.accountRegistration.enums.AccountRegistrationStatus;
import com.hedvig.paymentservice.domain.payments.DirectDebitStatus;
import com.hedvig.paymentservice.graphQl.types.BankAccount;
import com.hedvig.paymentservice.graphQl.types.PayinMethodStatus;
import com.hedvig.paymentservice.query.member.entities.Member;
import com.hedvig.paymentservice.query.member.entities.MemberRepository;
import com.hedvig.paymentservice.query.registerAccount.enteties.AccountRegistration;
import com.hedvig.paymentservice.query.registerAccount.enteties.AccountRegistrationRepository;
import com.hedvig.paymentservice.serviceIntergration.productPricing.ProductPricingService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.time.ZoneId;
import java.util.Comparator;
import java.util.Optional;

@Service
@Slf4j
public class BankAccountServiceImpl implements BankAccountService {

    private MemberRepository memberRepository;
    private AccountRegistrationRepository accountRegistrationRepository;
    private ProductPricingService productPricingService;

    @Autowired
    public BankAccountServiceImpl(MemberRepository memberRepository, AccountRegistrationRepository accountRegistrationRepository, ProductPricingService productPricingService) {
        this.memberRepository = memberRepository;
        this.accountRegistrationRepository = accountRegistrationRepository;
        this.productPricingService = productPricingService;
    }

    public BankAccount getBankAccount(String memberId) {

        if (memberId == null) {
            log.error("GetBankAccountInfo - hedvig.token is missing");
            throw new NullPointerException("GetBankAccountInfo - hedvig.token is missing");
        }
        Optional<Member> optionalMember = memberRepository.findById(memberId);

        return optionalMember.filter(x -> x.getDirectDebitStatus() == DirectDebitStatus.CONNECTED).map(BankAccount::fromMember).orElse(null);
    }

    public LocalDate getNextChargeDate(String memberId) {
        if (memberId == null) {
            log.error("GetNextChargeDate - hedvig.token is missing");
            throw new NullPointerException("GetNextChargeDate - hedvig.token is missing");
        }

        boolean hasContractActiveCurrentMonth = productPricingService.hasContractActiveCurrentMonth(memberId);
        if (!hasContractActiveCurrentMonth) {
            return null;
        }

        LocalDate today = LocalDate.now(ZoneId.of("Europe/Stockholm"));
        YearMonth currentPeriod = YearMonth.of(today.getYear(), today.getMonth());
        LocalDate chargeDateCurrentPeriod = getChargeDateOfPeriod(currentPeriod);
        if (!today.isAfter(chargeDateCurrentPeriod)) {
            return chargeDateCurrentPeriod;
        }
        return getChargeDateOfPeriod(currentPeriod.plusMonths(1));
    }

    //TODO: Handle red days
    static LocalDate getChargeDateOfPeriod(YearMonth period) {
        LocalDate chargeDateThisMonth = period.atDay(27);
        if (chargeDateThisMonth.getDayOfWeek() == DayOfWeek.SATURDAY) {
            chargeDateThisMonth = chargeDateThisMonth.plusDays(2);
        }
        if (chargeDateThisMonth.getDayOfWeek() == DayOfWeek.SUNDAY) {
            chargeDateThisMonth = chargeDateThisMonth.plusDays(1);
        }
        return chargeDateThisMonth;
    }

    public com.hedvig.paymentservice.graphQl.types.DirectDebitStatus getDirectDebitStatus(String memberId) {
        if (memberId == null) {
            log.error("GetDirectDebitStatus - hedvig.token is missing");
            throw new NullPointerException("GetDirectDebitStatus - hedvig.token is missing");
        }

        AccountRegistration accountRegistration = accountRegistrationRepository.
            findByMemberId(memberId).stream()
            .max(Comparator.comparing(AccountRegistration::getInitiated))
            .orElse(null);

        Optional<Member> optionalMember = memberRepository.findById(memberId);

        if (optionalMember.isPresent()) {
            Member member = optionalMember.get();
            if (member.getDirectDebitStatus() != null && member.getDirectDebitStatus().equals(DirectDebitStatus.CONNECTED)) {
                if (accountRegistration == null || accountRegistration.getStatus().equals(AccountRegistrationStatus.CONFIRMED) || accountRegistration.getStatus().equals(AccountRegistrationStatus.CANCELLED)) {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.ACTIVE;
                } else {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.PENDING;
                }
            } else if (member.getDirectDebitStatus() != null && member.getDirectDebitStatus().equals(DirectDebitStatus.DISCONNECTED)) {
                if (accountRegistration == null || accountRegistration.getStatus().equals(AccountRegistrationStatus.CONFIRMED) || accountRegistration.getStatus().equals(AccountRegistrationStatus.CANCELLED)) {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.NEEDS_SETUP;
                } else {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.PENDING;
                }
            } else if (member.getDirectDebitStatus() != null && member.getDirectDebitStatus().equals(DirectDebitStatus.PENDING)) {
                if (accountRegistration == null || accountRegistration.getStatus().equals(AccountRegistrationStatus.CANCELLED)) {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.NEEDS_SETUP;
                } else {
                    return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.PENDING;
                }
            }
        }
        return com.hedvig.paymentservice.graphQl.types.DirectDebitStatus.NEEDS_SETUP;
    }

    @Override
    public PayinMethodStatus getPayinMethodStatus(String memberId) {
        if (memberId == null) {
            log.error("getPayinMethodStatus - hedvig.token is missing");
            return PayinMethodStatus.NEEDS_SETUP;
        }
        Optional<Member> optionalMember = memberRepository.findById(memberId);

        if (optionalMember.isPresent()) {
            Member member = optionalMember.get();

            if (member.getTrustlyAccountNumber() != null) {
                return PayinMethodStatus.Companion.fromTrustlyDirectDebitStatus(getDirectDebitStatus(memberId));
            }

            if (member.getAdyenRecurringDetailReference() != null) {
                return member.getPayinMethodStatus();
            }

        }
        return PayinMethodStatus.NEEDS_SETUP;
    }
}

File: payment-service/src/main/java/com/hedvig/paymentservice/services/bankAccounts/BankAccountServiceImpl.kt
package com.hedvig.paymentservice.services.bankAccounts

import com.hedvig.paymentservice.domain.accountRegistration.enums.AccountRegistrationStatus
import com.hedvig.paymentservice.domain.payments.DirectDebitStatus
import com.hedvig.paymentservice.graphQl.types.BankAccount
import com.hedvig.paymentservice.graphQl.types.DirectDebitStatus as DirectDebitStatusDTO
import com.hedvig.paymentservice.graphQl.types.PayinMethodStatus
import com.hedvig.paymentservice.graphQl.types.PayinMethodStatus.Companion.fromAdyenAccountStatus
import com.hedvig.paymentservice.graphQl.types.PayinMethodStatus.Companion.fromTrustlyDirectDebitStatus
import com.hedvig.paymentservice.query.adyenAccount.MemberAdyenAccountRepository
import com.hedvig.paymentservice.query.directDebit.DirectDebitAccountOrder
import com.hedvig.paymentservice.query.directDebit.DirectDebitAccountOrderRepository
import com.hedvig.paymentservice.query.member.entities.MemberRepository
import com.hedvig.paymentservice.query.registerAccount.enteties.AccountRegistration
import com.hedvig.paymentservice.query.registerAccount.enteties.AccountRegistrationRepository
import com.hedvig.paymentservice.serviceIntergration.productPricing.ProductPricingService
import com.hedvig.paymentservice.util.getNextChargeChargeDate
import com.hedvig.paymentservice.web.dtos.DirectDebitAccountOrderDTO
import java.time.LocalDate
import org.springframework.stereotype.Service

@Service
class BankAccountServiceImpl(
    private val memberRepository: MemberRepository,
    private val accountRegistrationRepository: AccountRegistrationRepository,
    private val productPricingService: ProductPricingService,
    private val directDebitAccountOrderRepository: DirectDebitAccountOrderRepository,
    private val memberAdyenAccountRepository: MemberAdyenAccountRepository
) : BankAccountService {

    override fun getBankAccount(memberId: String): BankAccount? {
        val directDebitAccountOrder = getLatestDirectDebitAccountOrderEntity(memberId) ?: return null

        return BankAccount.fromDirectDebitAccountOrder(directDebitAccountOrder)
    }

    override fun getNextChargeDate(memberId: String): LocalDate? {
        val hasContractActiveCurrentMonth = productPricingService.hasContractActiveCurrentMonth(memberId)
        if (!hasContractActiveCurrentMonth) {
            return null
        }
        val market = productPricingService.getContractMarketInfo(memberId).market
        return getNextChargeChargeDate(market)
    }

    override fun getDirectDebitStatus(memberId: String): DirectDebitStatusDTO {
        val accountRegistration = accountRegistrationRepository
            .findByMemberId(memberId)
            .maxByOrNull { it.initiated }

        return when (getLatestDirectDebitAccountOrderEntity(memberId)?.directDebitStatus) {
            DirectDebitStatus.CONNECTED -> {
                if (accountRegistration.isNullOrConfirmedOrCancelled()) {
                    DirectDebitStatusDTO.ACTIVE
                } else {
                    DirectDebitStatusDTO.PENDING
                }
            }
            DirectDebitStatus.DISCONNECTED -> {
                if (accountRegistration.isNullOrConfirmedOrCancelled()) {
                    DirectDebitStatusDTO.NEEDS_SETUP
                } else {
                    DirectDebitStatusDTO.PENDING
                }
            }
            DirectDebitStatus.PENDING -> {
                if (accountRegistration.isNullOrCancelled()) {
                    DirectDebitStatusDTO.NEEDS_SETUP
                } else {
                    DirectDebitStatusDTO.PENDING
                }
            }
            null -> DirectDebitStatusDTO.NEEDS_SETUP
        }
    }

    override fun getPayinMethodStatus(memberId: String): PayinMethodStatus {
        val adyenAccountMaybe = memberAdyenAccountRepository.findById(memberId)

        if (adyenAccountMaybe.isPresent) {
            return fromAdyenAccountStatus(adyenAccountMaybe.get().accountStatus)
        }

        return fromTrustlyDirectDebitStatus(getDirectDebitStatus(memberId))
    }

    override fun getLatestDirectDebitAccountOrder(memberId: String): DirectDebitAccountOrderDTO? {
        val directDebitAccountOrder = getLatestDirectDebitAccountOrderEntity(memberId) ?: return null
        return DirectDebitAccountOrderDTO.from(directDebitAccountOrder)
    }

    private fun getLatestDirectDebitAccountOrderEntity(memberId: String): DirectDebitAccountOrder? {
        val directDebitAccountOrders = directDebitAccountOrderRepository.findAllByMemberId(memberId)
        return directDebitAccountOrders.maxByOrNull { it.createdAt }
    }

    private fun AccountRegistration?.isNullOrConfirmedOrCancelled(): Boolean = (
        this == null ||
            this.status == AccountRegistrationStatus.CONFIRMED ||
            this.status == AccountRegistrationStatus.CANCELLED
        )

    private fun AccountRegistration?.isNullOrCancelled(): Boolean = (
        this == null ||
            this.status == AccountRegistrationStatus.CANCELLED
        )
}
